### PR TITLE
ddl: add job version 2 and use typed args for truncate table

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #53930

Problem Summary:

### What changed and how does it work?
- add DDL job version 2 to use typed struct to replace old un-typed args slice
- add new typed args for TruncateTable
- currently we choose job version for new jobs randomly in test, see ddl.Start. Note: this only works for truncate-table for now.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

manually run TruncateTable tests using job version 2, success

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by joccau** [comment]
Maybe we could abstract some function, like this:

```
func setArgsForTruncateTable(job *model.job,  actualArgs *model.TruncateTableArgs, v version) {
    if v == 0 {
        // set args with the actualArgs.
        job.Args[0] = actualArg.NewTableID            
        job.Args[1] = ...
    } else {
       // set argsV2 with actualArgs.
        job.ArgsV2 = &model.TruncateTableArgs{
			NewTableID: 2,
		}
    }
}

func getArgsForTruncateTable(job *model.job, actualArgs *model.TruncateTableArgs, v version){
    if v == 0 {
        // set actualArgs with job.args
        newTableID = job.Args[0].(int64)
    } else {
       // set actualArgs with job.argsV2
       newTableID = job.ArgsV2.(*model.TruncateTableArgs).NewTableID
    }
}

func getIDCountForTruncateTable(job *model.job, v version) int  {
	if jobW.Version == model.JobVersion1 {
		partCount := jobW.Args[3].(int)
		count += 1 + partCount
	} else {
		count += 1 + len(jobW.ArgsV2.(*model.TruncateTableArgs).OldPartitionIDs)
	}
        return count.
}

```

And we can call this function in DDL jobSubmit and ddl jobScheduler and ....

**Comment by D3Hunter** [comment]
> Maybe we could abstract some function, like this:
> 
> ```
> func setArgsForTruncateTable(job *model.job,  actualArgs *model.TruncateTableArgs, v version) {
>     if v == 0 {
>         // set args with the actualArgs.
>         job.Args[0] = actualArg.NewTableID            
>         job.Args[1] = ...
>     } else {
>        // set argsV2 with actualArgs.
>         job.ArgsV2 = &model.TruncateTableArgs{
> 			NewTableID: 2,
> 		}
>     }
> }
> 
> func getArgsForTruncateTable(job *model.job, actualArgs *model.TruncateTableArgs, v version){
>     if v == 0 {
>         // set actualArgs with job.args
>         newTableID = job.Args[0].(int64)
>     } else {
>        // set actualArgs with job.argsV2
>        newTableID = job.ArgsV2.(*model.TruncateTableArgs).NewTableID
>     }
> }
> 
> func getIDCountForTruncateTable(job *model.job, v version) int  {
> 	if jobW.Version == model.JobVersion1 {
> 		partCount := jobW.Args[3].(int)
> 		count += 1 + partCount
> 	} else {
> 		count += 1 + len(jobW.ArgsV2.(*model.TruncateTableArgs).OldPartitionIDs)
> 	}
>         return count.
> }
> ```
> 
> And we can call this function in DDL jobSubmit and ddl jobScheduler and ...., put together functions that modify related parameters for different DDL type.

those part mostly are only used once in production code, and they mostly already inside a switch-case block which i think is clear enough to read, but yes the method became longer

for test code, i think it's acceptable now.

**Comment by D3Hunter** [comment]
some code does appear twice, i will try abstract them

File: pkg/ddl/executor.go [hunk 1]
```diff
@@ -6426,7 +6429,12 @@ func (e *executor) DoDDLJobWrapper(ctx sessionctx.Context, jobW *JobWrapper) (re
 func HandleLockTablesOnSuccessSubmit(ctx sessionctx.Context, jobW *JobWrapper) {
 	if jobW.Type == model.ActionTruncateTable {
 		if ok, lockTp := ctx.CheckTableLocked(jobW.TableID); ok {
-			newTableID := jobW.Args[0].(int64)
+			var newTableID int64
+			if jobW.Version == model.JobVersion1 {
```
Comment (lance6716) [inline pkg/ddl/executor.go:6433]: like `GetTruncateTableArgsBeforeRun`, this part can also be encapsulated?

File: pkg/ddl/executor.go [hunk 2]
```diff
@@ -6437,7 +6445,12 @@ func HandleLockTablesOnSuccessSubmit(ctx sessionctx.Context, jobW *JobWrapper) {
 func HandleLockTablesOnFinish(ctx sessionctx.Context, jobW *JobWrapper, ddlErr error) {
 	if jobW.Type == model.ActionTruncateTable {
 		if ddlErr != nil {
-			newTableID := jobW.Args[0].(int64)
+			var newTableID int64
+			if jobW.Version == model.JobVersion1 {
```
Comment (lance6716) [inline pkg/ddl/executor.go:6449]: ditto

File: pkg/ddl/job_scheduler.go [hunk 3]
```diff
@@ -723,7 +723,13 @@ func job2UniqueIDs(job *model.Job, schema bool) string {
 		if schema {
 			return strconv.FormatInt(job.SchemaID, 10)
 		}
-		return strconv.FormatInt(job.TableID, 10) + "," + strconv.FormatInt(job.Args[0].(int64), 10)
+		var newTableID int64
+		if job.Version == model.JobVersion1 {
```
Comment (lance6716) [inline pkg/ddl/job_scheduler.go:727]: ditto

File: pkg/ddl/job_submitter.go [hunk 4]
```diff
@@ -508,8 +519,12 @@ func getRequiredGIDCount(jobWs []*JobWrapper) int {
 			pInfo := jobW.Args[1].(*model.PartitionInfo)
 			count += len(pInfo.Definitions)
 		case model.ActionTruncateTable:
-			partCount := jobW.Args[3].(int)
-			count += 1 + partCount
+			if jobW.Version == model.JobVersion1 {
+				partCount := jobW.Args[3].(int)
```
Comment (lance6716) [inline pkg/ddl/job_submitter.go:523]: ditto

File: pkg/meta/model/job_args.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	job.ArgsV2 = v
+	return v, nil
+}
+
+// JobArgs is the interface for job arguments.
+type JobArgs interface {
+	// fillJob fills the job args for submitting job. we make it private to avoid
+	// calling it directly, use Job.FillArgs to fill the job args.
+	fillJob(job *Job)
```
Comment (joccau) [inline pkg/meta/model/job_args.go:41]: Rename `fillArgs` suggested.

Comment (D3Hunter) [reply pkg/meta/model/job_args.go:41]: from args's view, we are filling to job, see job.FillArgs

File: pkg/ddl/job_submitter.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/ddl/job_submitter.go:523]: this part only appear once

**Comment by D3Hunter** [comment]
@lance6716  also add another API for changed args, see https://github.com/pingcap/tidb/pull/55919/files#diff-0a328bfc50df4984d8b701f6cfb588c8bc675992e32e00ae05f2d900147ea8dfR44-R46

**Review by joccau — APPROVED** [review]

File: pkg/ddl/job_submitter.go [hunk 6]
```diff
@@ -579,13 +594,23 @@ func assignGIDsForJobs(jobWs []*JobWrapper, ids []int64) {
 			pInfo.NewTableID = pInfo.Definitions[0].ID
 		case model.ActionTruncateTable:
 			if !jobW.IDAllocated {
-				jobW.Args[0] = alloc.next()
-				partCount := jobW.Args[3].(int)
-				partIDs := make([]int64, partCount)
-				for i := range partIDs {
-					partIDs[i] = alloc.next()
+				if jobW.Version == model.JobVersion1 {
```
Comment (lance6716) [inline pkg/ddl/job_submitter.go:597]: for these getter logic, I prefer we always use a function to convert all version representations to `TruncateTableArgs`, then caller use `TruncateTableArgs`.

like
 
```golang
if !jobW.IDAllocated {
  arg := getTruncateTableArgs(job)
  arg.NewTableID = alloc.next()
  ...
```

**Review by lance6716 — APPROVED** [review]

File: pkg/ddl/job_submitter.go (hunk 6 above)
Comment (D3Hunter) [reply pkg/ddl/job_submitter.go:597]: for new jobs, i plan put JobArgs into JobWrapper, and fill to job on insert, so we can use it directly here.

**Review by tangenta — APPROVED** [review]

**Pull request merged**